[WIP] Add nullable annotation to Variant

quinmars opened the pull request:
Here the NRT raises some interesting questions: is it allowed to pass null values to the different constructor variants like `Variant(string? value)` or should we restrict it to `Variant(string value)`. The first version does not throw an can be serialized. The binary encoder will serialize it as an untyped null value. That's what the reference implementation does as well, so this seems to be correct. But if we should agree on this, shouldn't that also be true for `Variant(int[] value)`? Because here the constructor would throw on `null` values.

quinmars commented on UaClient/ServiceModel/Ua/Variant.cs:
```diff
@@ -344,8 +345,9 @@ public Variant(Enum value)
+            // REVIEW: Should we throw on null?
             this.Value = value;
             this.Type = VariantType.Boolean;
             this.ArrayDimensions = new int[value.Rank];
```
If we allow null values here as input, `value.Rank` would throw a NRE.

quinmars commented:
Please do not commit this PR before we discussed the questions above. I will then make all array constructors the same (nullable or non-nullable).

awcullen commented:
It would be useful to allow null array value in Variant constructor. Thanks for working through this.

On Dec 10, 2019, at 2:57 AM, Peter Wehrfritz <notifications@github.com<mailto:notifications@github.com>> wrote:


Please do not commit this PR before we discussed the questions above. I will then make all array constructors the same (nullable or non-nullable).

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/convertersystems/opc-ua-client/pull/121?email_source=notifications&email_token=AC2ZNGUY3ZUZRBOPGD64MHLQX5DWXA5CNFSM4JY2ESBKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGOJ7BI#issuecomment-563912581>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AC2ZNGWEJHJ627B7KBITBYTQX5DWXANCNFSM4JY2ESBA>.

quinmars commented:
Ok, I changed the behavior for `null` values. Not only for arrays but for all nullables. The reason is that there is no good way how you can describe a `default(string[])`. The type could be `VariantType.String`, but what should be the `ArrayDimension` for it? Same goes for `bool` or any other type. At the end this is not very important, because it will be encoded/decoded as a `VariantType.Null`. And the .NET world doesn't offer here a way to differ as well, `default(bool[]) == default(string)`. Hence I changed all constructors to create a null variant, i.e., with the type being `VariantType.Null` if the input value is `null`.
